### ISSUE TITLE
Notifications not working on Windows Vista (they used to work, but for some reason do not now)

### ISSUE THREAD

**Issue by casper** [bastimeyer/livestreamer-twitch-gui#305]
### Description

I am following several channels for which I have both checkmarks set in the channel settings ("Notifications" and "Receive notifications for this channel" are checked).

In the global settings I also have enabled desktop notifications.

But still no go. I get no notifications for any channel. 

It was a while since I used Livestreamer, but I seem to remember I had no problems getting the notifications before. Why they stopped working now I cannot seem to figure out. 

I've tried toggling the checkmarks on and off, rebooting and restarting but still no go :(

Is there anything else I could check or try to debug this further, or any idea why they would stop showing up?

### Environment details

LiveStreamer 0.14.2, Built 2016-08-08T17:03:10Z
NW.js 0.12.3

Windows Vista SP2 32bit

**Comment by bastimeyer** [comment]
> I had no problems getting the notifications before

Are you sure that you didn't misconfigure the notification settings and that the notification system is also not paused? See the context menu on the tray icon.
Notifications (in your case simple balloon tooltips since you're using Windows Vista 😵) can usually be disabled by your OS, maybe you should check this first.

Another reason could be missing read/write access to your system's temp folder, as *Livestreamer Twitch GUI* needs to download and cache channel icons for the notifications (balloon tooltips don't support icons, though).

> I have both checkmarks set in the channel settings 

You don't necessarily need to enable notifications for individual channels. This is only required if you've set the notification filter settings to "Ignore all except enabled ones". You're basically *whitelisting* notifications for channels that are *not blacklisted*.

> Is there anything else I could check or try to debug this further

No, but if you think this is caused by a bug, you're free to debug the code of the [NotificationService](https://github.com/bastimeyer/livestreamer-twitch-gui/blob/70b142626e9cf82585a20b89ea097a0552aeb1c6/src/app/services/NotificationService.js). Instructions can be found [here](https://github.com/bastimeyer/livestreamer-twitch-gui/blob/70b142626e9cf82585a20b89ea097a0552aeb1c6/CONTRIBUTING.md#developing-and-building).

> Windows Vista SP2 32bit

I really don't know why people are still using these kind of operating systems...
You should upgrade to a more up-to-date and secure OS. [Chromium doesn't support Vista anymore since almost a year ago.](https://chrome.googleblog.com/2015/11/updates-to-chrome-platform-support.html) This further means that once Livestreamer Twitch GUI will upgrade to the newer NW.js releases (which will hopefully be soon), Vista won't be supported anymore as well.

**Comment by casper** [comment]
Ok, now will follow some (probably) stupid questions. Sorry, I never used node before, but would like to debug the notifications to figure out what the problem is.

I installed node 4.5.0 + npm + downloaded the livestreamer gui source and checked out branch v0.14.2. I then ran the npm commands as instructed, which installed stuff, and then `grunt build`:

```
D:\Dev\livestreamer-twitch-gui>grunt build
Running "jshint:tasks" (jshint) task
>> 28 files lint free.

Running "jshint:app" (jshint) task
>> 255 files lint free.

Running "clean:tmp" (clean) task
>> 62 paths cleaned.

Running "copy:build" (copy) task
Created 1 directory, copied 17 files

Running "svgmin:flags" (svgmin) task
√ src/vendor/flag-icon-css/flags/4x3/bg.svg (saved 692 B 73
....
√ src/vendor/flag-icon-css/flags/4x3/vn.svg (saved 729 B 58
Total saved: 188.94 kB

Running "metadata:all" (metadata) task

Running "less:dev" (less) task
>> 1 stylesheet created.
>> 1 sourcemap created.

Running "requirejs:dev" (requirejs) task

Done.

D:\Dev\livestreamer-twitch-gui>
```

...and. What now? Nothing seems to happen. How do I start it? `grunt run` says:

```
Running "run:build" (run) task
Warning: No platform to build! Use --force to continue.

Aborted due to warnings.
```

Using `--force` doesn't help either. Is there something else I need to run still?

**Comment by bastimeyer** [comment]
> checked out branch v0.14.2

I linked you the build guides from the head of the master branch. Those are not compatible with the `v0.14.2` tag. Just checkout the master branch, reinstall the dependencies and run it again.

If you still want to debug `v0.14.2`, you need to run `grunt metadata less:source run:src`...

**Comment by casper** [comment]
Ok. I got it to run. Is there any way I can trigger a notification from the console to just do a quick check if they are working at all?

**Comment by casper** [comment]
Never mind. I'm going to close this ticket. I realize I have too little experience with node, and I would just keep asking you questions here to no end and waste your time. I found the Chrome extension Twitch Now which will have to do as a notification substitute for the moment. Thanks for your help.

**Comment by bastimeyer** [comment]
```js
var container = Ember.Application.NAMESPACES_BY_ID.App.__container__;
var Store = container.lookup( "service:store" );
var NotificationService = container.lookup( "service:notification" );
var stream = Store.peekAll( "twitchStream" ).slice( 0, 1 );
NotificationService.prepareNotifications( stream );
```
or
```js
var container = Ember.Application.NAMESPACES_BY_ID.App.__container__;
var Store = container.lookup( "service:store" );
var NotificationService = container.lookup( "service:notification" );
var stream = Store.peekAll( "twitchStream" ).slice( 0, 1 );
NotificationService.stripDisabledChannels( stream )
  .then( NotificationService.prepareNotifications.bind( NotificationService ) );
```

> I realize I have too little experience with node, and I would just keep asking you questions here to no end and waste your time.

Don't worry, it's fine...
Maybe you should check out the `nwjs-upgrade` or `node-notifier` branch (or simply see #275) if you don't want to debug the current code. The notifications have been rewritten there, so you could get *Chromium Rich Notifications* or *Growl Notifications* instead of balloon tooltips. There's also a "Test notification" button.
Those branches get force-pushed from time to time, though. Also, as I've said, Windows Vista is not being supported anymore, but you could still try it.

**Comment by casper** [comment]
How do I specify the platform when I run grunt? I keep getting `No platforms to build!` error.

**Comment by casper** [comment]
I hardcoded the platform into nw_build.js, so now I can run your console test code:

```
var container = Ember.Application.NAMESPACES_BY_ID.App.__container__;
var Store = container.lookup( "service:store" );
var NotificationService = container.lookup( "service:notification" );
var stream = Store.peekAll( "twitchStream" ).slice( 0, 1 );
NotificationService.prepareNotifications( stream );
```

And I get this in the console: 

```
Promise {[[PromiseStatus]]: "pending", [[PromiseValue]]: undefined}
```

But no notification shows up. So I guess we can conclude that the NW.js notification system just doesn't work...or?

I also looked if the channel icons are present in the `Temp` folder and readable by Chrome, and they are, so that's not the problem.

**Comment by bastimeyer** [comment]
> How do I specify the platform when I run grunt?

What does node say if you run `node -p "require('os').arch()"`? The build system expects either "x64" or "x86" (see [build/tasks/common/platforms.js](https://github.com/bastimeyer/livestreamer-twitch-gui/blob/master/build/tasks/common/platforms.js#L5)). If it doesn't match your arch, you can force it by running [`grunt release:win32`](https://github.com/bastimeyer/livestreamer-twitch-gui/blob/70b142626e9cf82585a20b89ea097a0552aeb1c6/CONTRIBUTING.md#building-and-compiling). See this [gitter chat log](https://gitter.im/bastimeyer/livestreamer-twitch-gui?at=5789a8133c5129720e367c44).

> So I guess we can conclude that the NW.js notification system just doesn't work

I don't know why it doesn't work on your system, I don't see any issues in my Windows VMs. The implementation of the NotificationService should also (most likely) not be the problem.

> And I get this in the console:

The notification system is completely asynchronous. What's being returned is just an instance of a `Promise`. You need to set some breakpoints to follow the flow.
I don't know which branch you are on, but [`showNotification`](https://github.com/bastimeyer/livestreamer-twitch-gui/blob/70b142626e9cf82585a20b89ea097a0552aeb1c6/src/app/services/NotificationService.js#L435) needs to be called.

**To make it all a bit simpler:**
Run the application from the master branch, open the console and paste `new window.Notification("Test")`. This is the old and simple NW.js API. If this doesn't work, you've either disabled notifications somewhere in your system or NW.js is failing.

As I've said, the notifications are being reworked right now, so you might want to test the new ones instead from the `node-notifier` branch (although not finished). Deleting your config folder located at `%LOCALAPPDATA%\livestreamer-twitch-gui` could help too, since I don't know if your settings are correct, the default ones are for sure.

**Comment by casper** [comment]
> What does node say if you run node -p "require('os').arch()"?

`ia32`

> I don't know which branch you are on, but showNotification needs to be called.

I'm on the master branch. I don't know how to call `showNotification`, but I assume you mean the system needs to call it after the notification has been queued. But of course I don't know how to make that happen either :-/ Lol...

> To make it all a bit simpler 

This headline I like. Alas..running `new window.Notification("Test")` produces no effect.
I also deleted the config folder but it didn't help either.

Next I switched to  `node-notifier`, but wondering if I need to clean the build somehow after I switch branches? Is it enough to just run `npm install` again and then `grunt build` will do the right thing?

I'm asking because running `grunt build` on this new branch now, after it finishes building, it says `Launching App` like before, but the Livestreamer GUI window never shows up (I see `nw.exe` in process explorer though).

**Comment by bastimeyer** [comment]
> ia32

Fixed in 22a6349c2e57f1c0e69ca5e02cfcfd39c3efeeba

> running new window.Notification("Test") produces no effect.

This is the basic Notification-API in NW.js. If this doesn't show anything, then it's a NW.js issue on your system or notifications (balloon tooltips) are disabled in Windows itself. I don't use Windows, but if I remember correctly it's the same as in Win7 and balloon tooltips can be enabled/disabled in the tray bar options for each application individually.

> Is it enough to just run npm install again and then grunt build will do the right thing?

Yes.

> but the Livestreamer GUI window never shows up

Did you kill all running NW.js processes first? The application can only be run in single instance mode and the builder doesn't know if a process is already running.
Also, as I've already said, the new NW.js releases use a newer Chromium version that doesn't support Vista anymore, so don't be surprised if doesn't work at all.

**Comment by casper** [comment]
> Fixed in 22a6349

Thx. Works now.

> Did you kill all running NW.js processes first?

Yes. I checked and the newer NW.js uses Chromium 50+, when 49 is the last one for Vista. So yes..that seems to be the reason.

I also found a small Windows command line utility called `notifu` which can generate balloon tips for batch file use etc., and it does not work on my computer either. So there we have it. Wild goose chase ends.

I still have no clue why balloon tips are broken. You can only disable them in Vista with a registry hack, and I checked this registry key already before I started debugging. Very very strange. Problem is I can't upgrade this laptop for many reasons, but ok...at least we found the cause now. Thanks a lot for your help! Now I know how to debug Node apps at least...

**Comment by casper** [comment]
Fix for disabled balloon notifications in Vista: Start Button / Properties / Notification Area / Customize / Default Settings / Ok / Ok. Pressing default settings (+ Ok two times, otherwise does not work) seems to reset something which re-enables notifications. Everything works now.
